File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/telegram/MessageEntityArrayFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.telegram;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;
import ru.ioleksiv.telegram.bot.api.model.objects.std.MessageEntity;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface MessageEntityArrayFilter {

    AnnotationState status() default AnnotationState.ON;

    MessageEntity.Type type() default MessageEntity.Type.ALL;

    IntegerFilter length() default @IntegerFilter(status = AnnotationState.OFF);

    StringFilter url() default @StringFilter(status = AnnotationState.OFF);

    UserFilter user() default @UserFilter(status = AnnotationState.OFF);

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/api/annotations/filter/composite/MessageEntityArrayFilter.kt
package io.github.oleksivio.telegram.bot.api.annotations.filter.composite

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.IntegerFilter
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter
import io.github.oleksivio.telegram.bot.api.model.annotation.AnnotationState
import io.github.oleksivio.tl.kbot.server.api.objects.std.MessageEntity

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class MessageEntityArrayFilter(
    val status: AnnotationState = AnnotationState.ON,
    val type: MessageEntity.Type = MessageEntity.Type.UNKNOWN,
    val validator: Array<String> = [],
    val length: IntegerFilter = IntegerFilter(status = AnnotationState.OFF),
    val url: StringFilter = StringFilter(status = AnnotationState.OFF),
    val user: UserFilter = UserFilter(status = AnnotationState.OFF)
)
